yarl/_helpers.pyx

```diff
@@ -0,0 +1,35 @@
+cdef class cached_property:
+    """Use as a class method decorator.  It operates almost exactly like
+    the Python `@property` decorator, but it puts the result of the
+    method it decorates into the instance dict after the first call,
+    effectively replacing the function it decorates with an instance
+    variable.  It is, in Python parlance, a data descriptor.
+
+    """
+
+    cdef object wrapped
+    cdef object name
+
+    def __init__(self, wrapped):
+        self.wrapped = wrapped
+        self.name = wrapped.__name__
+
+    @property
+    def __doc__(self):
+        return self.wrapped.__doc__
+
+    def __get__(self, inst, owner):
+        try:
+            try:
+                return inst._cache[self.name]
```

Comment:
In a future PR we might be able to improve this a bit more as it could use the PyDict functions

<img width="1120" alt="Screenshot 2024-08-31 at 9 09 25 AM" src="https://github.com/user-attachments/assets/b7049ed4-c696-486a-8788-d8e15fc55a6f">